shared/js/dataDriver.js

```diff
@@ -305,15 +316,27 @@ loop.DataDriver = function() {
      *  - {Mixed}  value     Stored value for the record.
      */
     _processRecord(key, data) {
+      let dispatchAction, dispatchObject;
       // XXX akita bug 1274130: Validate data/key/timestamp before processing.
-      let [, type] = key.match(/^([^!]+)!(.+)$/);
+      let [, type, id] = key.match(/^([^!]+)!(.+)$/);
       switch (type) {
-        case "chat": {
-          let chat = data.value;
-          chat.receivedTimestamp = new Date(data.timestamp).toISOString();
-          this._dispatcher.dispatch(new actions.ReceivedTextChatMessage(chat));
+        case "chat":
+          dispatchAction = "ReceivedTextChatMessage";
+          dispatchObject = Object.assign(data.value, {
+            receivedTimestamp: new Date(data.timestamp).toISOString()
+          });
```

Comment:
I'm not a huge fan of writing to formal parameters -- it's error prone, since JavaScript passes a copy of object references.  In this case, I think it means that the changes being written into data.value for local use will also be visible to the caller after this returns.  If that's actually how this is behaving, I'd suggest changing the behavior to not write data.value.  If that's important for some reason, however, at the very least it wants to be documented in the JSDoc that this method has intentional side effects.